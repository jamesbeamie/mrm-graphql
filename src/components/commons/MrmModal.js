/* eslint-disable no-underscore-dangle */
/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ActionButton from './ActionButtons';
import IconButtons from './IconButtons';
import { getUserDetails } from '../helpers/QueriesHelpers';
import '../../assets/styles/mrmmodal.scss';

class MrmModal extends Component {
  static propTypes = {
    modalContent: PropTypes.node,
    title: PropTypes.string,
    buttonText: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    btnImage: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    type: PropTypes.number,
    withButton: PropTypes.bool,
    styleClassName: PropTypes.string,
    handleCloseModal: PropTypes.func,
    showActionButton: PropTypes.bool,
  };

  static defaultProps = {
    modalContent: '',
    buttonText: '',
    title: '',
    btnImage: '',
    type: 1,
    withButton: true,
    styleClassName: '',
    handleCloseModal: () => {},
    showActionButton: true,
  };

  state = {
    isOpen: false,
    role: '',
  };

  componentDidMount() {
    this._isMounted = true;
    this.props.withButton && this.setUserRole();
  }

  componentWillUnmount() {
    this._isMounted = false;
  }
  /**
   * Sets the role of the current user
   * @returns {void}
   */
  setUserRole = async () => {
    const user = await getUserDetails();
    this._isMounted && this.setState({ role: user.roles[0].id });
  };

  _isMounted = false;
  /**
   * Toggle the visibility of the modal
   *
   * @returns {void}
   */
  toggleModal = () => {
    this.setState(({ isOpen }) => {
      if (isOpen) this.props.handleCloseModal();
      return { isOpen: !this.state.isOpen };
    });
  };

  toggleModalByRef = () => {
    this._isMounted && this.setState({ isOpen: !this.state.isOpen }, () => {});
  };

  render() {
    const {
      actionButtonText,
      handleSubmit,
      cancelButtonText,
      buttonText,
      title,
      type,
      modalContent,
      isLoading,
      iconButtonClass,
      withButton,
      styleClassName,
      showActionButton,
      btnImage,
    } = this.props;
    return (
      <div className="modal-component">
        {type === 1 &&
          this.state.role === '2' && (
            <IconButtons
              btnImage={btnImage}
              buttonText={buttonText}
              openModal={this.toggleModal}
              modalButtonClassName={iconButtonClass}
            />
          )}
        {this.state.isOpen && (
          <div className="modal">
            <div className="overlay" />
            <div className={`modal_content ${styleClassName}`}>
              <h2>{title}</h2>
              {modalContent}
              {withButton && (
                <ActionButton
                  cancelButtonText={cancelButtonText}
                  actionButtonText={actionButtonText}
                  onClickSubmit={handleSubmit}
                  onClickCancel={this.toggleModal}
                  isLoading={isLoading}
                  showActionButton={showActionButton}
                />
              )}
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default MrmModal;
