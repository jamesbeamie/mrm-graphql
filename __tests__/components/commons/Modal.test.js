import React from 'react';
import { mount } from 'enzyme';
import MrmModal from '../../../src/components/commons/Modal';

describe('MrmModal Component', () => {
  const wrapper = mount(<MrmModal
    title="hello"
    buttonText="button"
    handleCloseRequest={jest.fn()}
    closeModal
    children={<div className="test-div">Hello</div>}
  />);

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have one button', () => {
    const modalForm = wrapper.find('button');
    expect(modalForm).toHaveLength(1);
    expect(modalForm.text()).toEqual('button');
  });

  it('handles openModal', () => {
    expect(wrapper.instance().openModal());
  });

  it('handles closeModal()', () => {
    expect(wrapper.instance().closeModal());
  });

  it('handles componentDidUpdate()', () => {
    expect(wrapper.instance().componentDidUpdate('foo', {}, jest.fn()));
  });
});
