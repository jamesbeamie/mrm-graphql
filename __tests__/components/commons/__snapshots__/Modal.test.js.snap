// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MrmModal Component renders properly 1`] = `
<MrmModal
  buttonText="button"
  className="modalClass"
  closeModal={true}
  handleCloseRequest={[MockFunction]}
  modalButtonClassName="button"
  openModal={false}
  title="hello"
>
  <IconButtons
    btnImage=""
    buttonText="button"
    disabled={false}
    modalButtonClassName="button"
    openModal={[Function]}
  >
    <button
      className="button"
      disabled={false}
      id="modal-button"
      onClick={[Function]}
    >
      button
    </button>
  </IconButtons>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    className="modalClass"
    closeTimeoutMS={0}
    contentLabel="Parent Modal"
    isOpen={false}
    onAfterOpen={[Function]}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "backgroundColor": "#ffffff",
          "border": "1px solid #CCCCCC",
          "borderRadius": "4px",
          "bottom": "auto",
          "boxShadow": "0 2px 40px 10px rgba(185, 180, 180, 0.2)",
          "fontFamily": "DINPro-Regular",
          "left": "50%",
          "outline": "none",
          "overflow": "auto",
          "position": "absolute",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
      }
    }
  >
    <Portal
      containerInfo={
        <div
          class="ReactModalPortal"
        />
      }
    >
      <ModalPortal
        ariaHideApp={false}
        bodyOpenClassName="ReactModal__Body--open"
        className="modalClass"
        closeTimeoutMS={0}
        contentLabel="Parent Modal"
        defaultStyles={
          Object {
            "content": Object {
              "WebkitOverflowScrolling": "touch",
              "background": "#fff",
              "border": "1px solid #ccc",
              "borderRadius": "4px",
              "bottom": "40px",
              "left": "40px",
              "outline": "none",
              "overflow": "auto",
              "padding": "20px",
              "position": "absolute",
              "right": "40px",
              "top": "40px",
            },
            "overlay": Object {
              "backgroundColor": "rgba(255, 255, 255, 0.75)",
              "bottom": 0,
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
            },
          }
        }
        isOpen={false}
        onAfterOpen={[Function]}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "backgroundColor": "#ffffff",
              "border": "1px solid #CCCCCC",
              "borderRadius": "4px",
              "bottom": "auto",
              "boxShadow": "0 2px 40px 10px rgba(185, 180, 180, 0.2)",
              "fontFamily": "DINPro-Regular",
              "left": "50%",
              "outline": "none",
              "overflow": "auto",
              "position": "absolute",
              "right": "auto",
              "top": "50%",
              "transform": "translate(-50%, -50%)",
            },
          }
        }
      />
    </Portal>
  </Modal>
</MrmModal>
`;
